pkg/spdk/server.go

```diff
@@ -1502,3 +1584,218 @@ func (s *Server) VersionDetailGet(context.Context, *emptypb.Empty) (*spdkrpc.Ver
+	}
+
+	// Don't recreate the backing image
+	backingImageSnapLvolName := GetBackingImageSnapLvolName(req.Name, req.LvsUuid)
+	if _, ok := s.backingImageMap[backingImageSnapLvolName]; ok {
+		return nil, grpcstatus.Errorf(grpccodes.AlreadyExists, "backing image %v already exists", req.Name)
+	}
+
+	bi, err := s.newBackingImage(req)
+	if err != nil {
+		return nil, err
+	}
+
+	s.RLock()
+	spdkClient := s.spdkClient
+	s.RUnlock()
+
+	return bi.Create(spdkClient, s.portAllocator, req.FromAddress, req.SrcLvsUuid)
+}
+
+func (s *Server) BackingImageDelete(ctx context.Context, req *spdkrpc.BackingImageDeleteRequest) (ret *emptypb.Empty, err error) {
+	if req.Name == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "backing image name is required")
+	}
+	if req.LvsUuid == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "lvs UUID is required")
+	}
+
+	s.RLock()
+	bi := s.backingImageMap[GetBackingImageSnapLvolName(req.Name, req.LvsUuid)]
+	spdkClient := s.spdkClient
+	s.RUnlock()
+
+	defer func() {
+		if err == nil {
+			s.Lock()
+			delete(s.backingImageMap, GetBackingImageSnapLvolName(req.Name, req.LvsUuid))
+			s.Unlock()
+		}
+	}()
+
+	if bi != nil {
+		if err := bi.Delete(spdkClient, s.portAllocator); err != nil {
+			return nil, grpcstatus.Error(grpccodes.Internal, errors.Wrapf(err, "failed to delete backing image %v in lvs %v", req.Name, req.LvsUuid).Error())
+		}
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+func (s *Server) BackingImageGet(ctx context.Context, req *spdkrpc.BackingImageGetRequest) (ret *spdkrpc.BackingImage, err error) {
+	if req.Name == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "backing image name is required")
+	}
+	if req.LvsUuid == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "lvs UUID is required")
+	}
+
+	backingImageSnapLvolName := GetBackingImageSnapLvolName(req.Name, req.LvsUuid)
+
+	s.RLock()
+	bi := s.backingImageMap[backingImageSnapLvolName]
+	s.RUnlock()
+
+	if bi == nil {
+		lvsName := GetLvsNameByUUID(s.spdkClient, req.LvsUuid)
+		if lvsName != "" {
+			backingImageSnapLvolAlias := spdktypes.GetLvolAlias(lvsName, backingImageSnapLvolName)
+			bdevLvolList, err := s.spdkClient.BdevLvolGet(backingImageSnapLvolAlias, 0)
+			if err != nil {
+				return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find backing image %v in lvs %v", req.Name, req.LvsUuid)
+			}
+			if len(bdevLvolList) != 1 {
+				return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find backing image %v in lvs %v", req.Name, req.LvsUuid)
+			}
+			// If we can get the lvol, verify() will reconstruct the backing image record in the server, should inform the caller
+			return nil, grpcstatus.Errorf(grpccodes.NotFound, "backing image %v lvol found in the lvs %v but failed to find the record in the server", req.Name, req.LvsUuid)
+		}
+		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find backing image %v in lvs %v", req.Name, req.LvsUuid)
+	}
```

Comment:
Do we really need to check the lvs and the lvol list? Is it for providing more information for troubleshooting?

Reply:
fixed the error message